### PR TITLE
Strip GCC comments from output before parsing in get_define. Closes #…

### PR DESCRIPTION
**Pull request by jpakkane**
…1726.

No test, because I don't want to add build-dep on cups just for this. If someone can come up with a way to create the problematic output, please post below.

### CODE REVIEW

**Comment by nirbheek** [comment]
This is very interesting. Reproduceable test case:

```c
#ifndef CUPS_VERSION_MAJOR
# define CUPS_VERSION_MAJOR
#endif
#include <cups/cups.h>
MESON_GET_DEFINE_DELIMITER
CUPS_VERSION_MAJOR
```
Compile as: `gcc -E test.c`. I will try to figure out what's causing this and remove the dependency on cups.

**Comment by jpakkane** [comment]
If you get a fix, just push it to this branch.

**Comment by tp-m** [comment]
On a side note, shouldn't the `#ifndef` `#define` bit be *after* the prefix include?

**Comment by nirbheek** [comment]
> On a side note, shouldn't the #ifndef #define bit be after the prefix include?

Yes, indeed they should. Thinko by me while writing that code.

**Comment by nirbheek** [comment]
I've added a test that exposes the bug locally for me:

```
Project name: get define
Native c compiler: ccache cc (gcc 6.3.1)
Native cpp compiler: ccache c++ (gcc 6.3.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Checking for value of define "__linux__": 1
Library z found: YES
Checking for value of define "ZLIB_VER_MAJOR": 
# 7 "/tmp/tmpmarut4_2/testfile.c" 3 4
1

Meson encountered an error in file meson.build, line 29, column 4:
Assert failed: ZLIB_VER_MAJOR value is "
# 7 "/tmp/tmpmarut4_2/testfile.c" 3 4
1" instead of "1"
```

This is a really weird gcc edge-case. When a C file containing `#include <foo.h>` is pre-processed and `foo.h` is found in the compiler's default search path, GCC inserts an extra comment between the delimiter and the define which causes a parsing error. AFAICT there's no GCC option to turn this line-number commenting of preprocessed output off.

File: mesonbuild/compilers.py [hunk 1]
```diff
@@ -1151,7 +1151,13 @@ def get_define(self, dname, prefix, env, extra_args, dependencies):
                 raise EnvironmentException('Could not get define {!r}'.format(dname))
         # Get the preprocessed value after the delimiter,
         # minus the extra newline at the end
-        return p.stdo.split(delim + '\n')[-1][:-1]
+        valuestr = p.stdo.split(delim + '\n')[-1]
+        # Sometimes GCC adds comment lines, ignore them.
+        for line in valuestr.split('\n'):
+            if len(line) > 0 and line[0] == '#':
+                continue
+            return line
```
Comment (nirbheek) [inline mesonbuild/compilers.py]: This will only return a single line of output while the earlier implementation returned multiline defines too.

Comment (nirbheek) [inline mesonbuild/compilers.py]: Actually, this returns an empty string in all cases that contain the `#` now.

**Comment by nirbheek** [comment]
> AFAICT there's no GCC option to turn this line-number commenting of preprocessed output off.

Obviously I am stupid and wrong, because as soon as I found the option, I realized that I had used it multiple times in the past...

I've added a patch that fixes this in a much more concise and always-correct way now + tests.

**Comment by jpakkane** [comment]
Good stuff, thanks.

**Pull request merged**